File: generate_projects.js
const fs = require('fs');
const path = require('path');

const rootDir = 'd:\\Projects\\Portfolio\\Portfolio';
const projectsDir = path.join(rootDir, 'assets', 'projects');

// Helper to find all project_info.txt files
function findProjectFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    files.forEach(file => {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);
        if (stat.isDirectory()) {
            // Skip the weird '0' directory in 'öbb' if it contains a project_info.txt that we want to ignore
            // But we actually walk everything and filter later based on content
            findProjectFiles(filePath, fileList);
        } else {
            if (file === 'project_info.txt') {
                fileList.push(filePath);
            }
        }
    });
    return fileList;
}

const projectFiles = findProjectFiles(projectsDir);

const projects = [];

// Helper to check if JSON exists for a directory
function getJsonPath(txtPath) {
    const dir = path.dirname(txtPath);
    const jsonPath = path.join(dir, 'project_info.json');
    if (fs.existsSync(jsonPath)) {
        return jsonPath;
    }
    return null;
}

// We might have duplicates if we list both txt and json, or if we stick to listing folders.
// The current findProjectFiles finds 'project_info.txt'.
// Let's change how we iterate. We have a list of txt files.
// For each txt file, check if there is a sibling json file.
// If yes, use JSON and SKIP the txt parsing.
// If no, parse txt.

projectFiles.forEach(txtFilePath => {
    const jsonFilePath = getJsonPath(txtFilePath);

    if (jsonFilePath) {
        // Parse JSON
        const content = fs.readFileSync(jsonFilePath, 'utf-8');
        try {
            const data = JSON.parse(content);

            // Resolve Images in Blocks
            // In JSON, user might specify "1.jpg". We need to resolve to relative path "assets/projects/..."
            const dir = path.dirname(jsonFilePath);

            const resolveImg = (imgName) => {
                if (!imgName) return "";
                const absolutePath = path.join(dir, imgName);
                if (fs.existsSync(absolutePath)) {
                    let relPath = path.relative(rootDir, absolutePath);
                    return relPath.replace(/\\/g, '/');
                }
                return ""; // Or placeholder
            };

            const heroImage = resolveImg(data.heroImage);

            const blocks = (data.blocks || []).map(b => {
                return {
                    type: b.type,
                    image: resolveImg(b.image),
                    text: b.text
                };
            });

            projects.push({
                id: data.id,
                originalName: data.projectName,
                title: data.title,
                category: data.category,
                status: data.status,
                year: data.year, // JSON uses "year" or "lines" uses "Key Dates"
                tools: data.technologies, // JSON "technologies", txt "Technologies"
                heroImage: heroImage,
                description: data.description,
                role: { en: "Designer", de: "Designer" }, // Could add to JSON
                blocks: blocks,
                isMain: jsonFilePath.includes('assets\\projects\\main')
            });

        } catch (e) {
            console.error("Error parsing JSON:", jsonFilePath, e);
        }

    } else {
        // Parse TXT (Legacy Logic)
        const filePath = txtFilePath;
        const content = fs.readFileSync(filePath, 'utf-8');

        // Parse Key-Values
        const lines = content.split(/\r?\n/);
        const data = {};
        let currentKey = null;
        let descriptionBuffer = [];
        let isDescription = false;

        lines.forEach(line => {
            const match = line.match(/^([A-Za-z ]+): ?(.*)$/);
            if (match && !isDescription) {
                const key = match[1].trim();
                const value = match[2].trim();
                if (key === 'Description') {
                    isDescription = true;
                    data[key] = value; // First line of description
                } else {
                    data[key] = value;
                }
            } else if (isDescription) {
                descriptionBuffer.push(line);
            }
        });

        // Check if ID is valid
        if (!data['ID'] || data['ID'].trim() === '') {
            // Skip files without valid ID (like the one in öbb/0)
            return;
        }

        // Process Description
        const fullDescription = (data['Description'] || '') + '\n' + descriptionBuffer.join('\n');

        // Split by " - " (isolated dashes)
        // Regex to match - on its own line or surrounded by spaces
        const parts = fullDescription.split(/\n\s*-\s*\n/);

        // Clean parts
        const textBlocks = parts.map(p => p.trim()).filter(p => p.length > 0);

        // Find Images in the directory
        const dir = path.dirname(filePath);
        const files = fs.readdirSync(dir);

        // Helper to find image by number
        const findImage = (num) => {
            const exts = ['.jpg', '.png', '.jpeg', '.gif', '.webp'];
            for (const ext of exts) {
                const name = `${num}${ext}`;
                if (files.includes(name)) {
                    // Return relative path for HTML
                    // Path from d:\Projects\Portfolio\Portfolio\index.html to image
                    // Image path: dir + name
                    // Relative: Path.relative(rootDir, dir) + / + name
                    let relPath = path.relative(rootDir, path.join(dir, name));
                    return relPath.replace(/\\/g, '/');
                }
            }
            return null;
        };

        const heroImage = findImage(0); // Image 0

        let actualHero = heroImage;
        if (!actualHero) {
            // fallback
            const firstImg = findImage(1);
            if (firstImg) actualHero = firstImg;
        }

        // Build Blocks
        // Verify how many images exist
        let maxImgNum = 0;
        while (true) {
            if (findImage(maxImgNum + 1)) {
                maxImgNum++;
            } else {
                break;
            }
        }

        const maxIndex = Math.max(textBlocks.length - 1, maxImgNum);

        const blocks = [];
        for (let i = 1; i <= maxIndex; i++) {
            const text = textBlocks[i] || "";
            const img = findImage(i);

            let block = {};
            if (img) {
                block = {
                    type: 'image-text',
                    image: img,
                    text: {
                        en: text, de: text
                    }
                };
            } else if (text) {
                block = {
                    type: 'text-only',
                    text: {
                        en: text, de: text
                    }
                };
            } else {
                continue;
            }
            blocks.push(block);
        }

        projects.push({
            id: data['ID'],
            originalName: data['Project Name'],
            title: { en: data['Project Name'], de: data['Project Name'] },
            category: { en: data['Category'], de: data['Category'] },
            status: data['Status'],
            year: data['Key Dates'],
            tools: data['Technologies'],
            heroImage: actualHero,
            description: { en: textBlocks[0], de: textBlocks[0] }, // First block handles main description
            role: { en: "Designer", de: "Designer" }, // Role not always in file? 
            blocks: blocks,
            // For sorting
            isMain: filePath.includes('assets\\projects\\main')
        });
    }
});

// Sort Projects
// Move Main projects to top?
// And reassign IDs 1-N
projects.sort((a, b) => {
    if (a.isMain && !b.isMain) return -1;
    if (!a.isMain && b.isMain) return 1;
    return a.originalName.localeCompare(b.originalName);
});

// Re-assign IDs
projects.forEach((p, i) => {
    p.newId = i + 1;
});

// Generate JSON
const jsonOutput = {};
projects.forEach(p => {
    jsonOutput[p.newId] = {
        title: p.title,
        category: p.category,
        heroImage: p.heroImage || "",
        description: p.description,
        role: p.role,
        year: p.year,
        tools: p.tools,
        blocks: p.blocks
    };
});

console.log("JSON OBJECT START");
console.log(JSON.stringify(jsonOutput, null, 2));
console.log("JSON OBJECT END");

// Generate HTML
// We need to generate the grid items.
// Using template based on index.html
/*
<div class="reveal-on-scroll group cursor-pointer" onclick="openModal(${newId})">
    <div class="hover-target relative aspect-[4/3] bg-gray-100 overflow-hidden mb-4">
        <img src="${heroImage}"
            class="w-full h-full object-cover transition-transform duration-700 group-hover:scale-105"
            alt="${title}">
        <div class="absolute inset-0 bg-black/0 group-hover:bg-black/10 transition-colors duration-300"></div>
    </div>
    <div class="flex justify-between items-start">
        <div>
            <h3 class="text-lg font-semibold">${title}</h3>
            <p class="text-sm text-gray-500">${category}</p>
        </div>
    </div>
</div>
*/

let visibleHtml = "";
let hiddenHtml = "";

projects.forEach((p, i) => {
    const html = `
    <!-- Project ${p.newId} -->
    <div class="reveal-on-scroll group cursor-pointer ${i >= 3 ? '' : ''}" onclick="openModal(${p.newId})">
        <div class="hover-target relative aspect-[4/3] bg-gray-100 overflow-hidden mb-4">
            <img src="${p.heroImage || 'https://via.placeholder.com/800x600?text=No+Image'}"
                class="w-full h-full object-cover transition-transform duration-700 group-hover:scale-105"
                alt="${p.title.en}"
                onerror="this.src='https://via.placeholder.com/800x600?text=Image+Unavailable'">
            <div
                class="absolute inset-0 bg-black/0 group-hover:bg-black/10 transition-colors duration-300">
            </div>
        </div>
        <div class="flex justify-between items-start">
            <div>
                <h3 class="text-lg font-semibold">${p.title.en}</h3>
                <p class="text-sm text-gray-500">${p.category.en}</p>
            </div>
        </div>
    </div>`

    if (i < 3) {
        visibleHtml += html;
    } else {
        hiddenHtml += html;
    }
});

console.log("HTML VISIBLE START");
console.log(visibleHtml);
console.log("HTML VISIBLE END");

console.log("HTML HIDDEN START");
console.log(hiddenHtml);
console.log("HTML HIDDEN END");

File: generate_projects.py
import os
import re
import json

root_dir = r"d:\Projects\Portfolio\Portfolio"
projects_dir = os.path.join(root_dir, "assets", "projects")

project_list = []

def find_project_files(dir_path):
    files = []
    for root, dirnames, filenames in os.walk(dir_path):
        if "project_info.txt" in filenames:
            if os.path.basename(root) == '0' and 'öbb' in root:
                 continue
            files.append(os.path.join(root, "project_info.txt"))
    return files

project_files = find_project_files(projects_dir)

def find_image(root, num):
    try:
        image_files = [f for f in os.listdir(root) if f.lower().endswith(('.jpg', '.png', '.jpeg', '.gif', '.webp', '.mp4'))]
    except OSError:
        return None

    for ext in ['.jpg', '.png', '.jpeg', '.gif', '.webp', '.mp4']:
        name = f"{num}{ext}"
        if name in image_files:
            path_rel = os.path.relpath(os.path.join(root, name), root_dir).replace("\\", "/")
            return path_rel

    for suffix in ['a', 'b', 'c', 'd', 'e']:
        for ext in ['.jpg', '.png', '.jpeg', '.gif', '.webp', '.mp4']:
            name = f"{num}{suffix}{ext}"
            for f in image_files:
                if f.lower() == name.lower():
                     path_rel = os.path.relpath(os.path.join(root, f), root_dir).replace("\\", "/")
                     return path_rel
    return None

for file_path in project_files:
    root = os.path.dirname(file_path)
    content = ""
    try:
        with open(file_path, "r", encoding="utf-8-sig") as f:
            content = f.read()
    except UnicodeDecodeError:
        try:
            with open(file_path, "r", encoding="latin-1") as f:
                content = f.read()
        except:
            continue

    data = {}
    lines = content.splitlines()
    is_description = False
    description_lines = []
    
    for line in lines:
        match = re.match(r"^([A-Za-z ]+): ?(.*)$", line)
        if match and not is_description:
            key = match.group(1).strip()
            val = match.group(2).strip()
            if key == "Description":
                is_description = True
                data[key] = val
            else:
                data[key] = val
        elif is_description:
            description_lines.append(line)
    
    if "ID" not in data or not data["ID"]:
        continue
        
    full_desc = (data.get("Description", "") + "\n" + "\n".join(description_lines)).strip()
    parts = re.split(r'\n\s*-\s*\n', full_desc)
    blocks_text = [p.strip() for p in parts if p.strip()]
    
    hero_image = find_image(root, 0)
    display_hero = hero_image if hero_image else find_image(root, 1)
    
    project_blocks = []
    main_desc = blocks_text[0] if blocks_text else ""
    
    for idx, text in enumerate(blocks_text):
        if idx == 0: continue
        img_path = find_image(root, idx)
        block = {}
        if img_path:
            block = {
                "type": "image-text",
                "image": img_path, 
                "text": {"en": text, "de": text}
            }
        else:
            block = {
                "type": "text-only",
                "text": {"en": text, "de": text}
            }
        project_blocks.append(block)
        
    project_list.append({
        "original_id": data.get("ID"),
        "original_name": data.get("Project Name"),
        "category": data.get("Category", "Design"),
        "status": data.get("Status", ""),
        "year": data.get("Key Dates", ""),
        "tools": data.get("Technologies", ""),
        "hero_image": display_hero,
        "description": main_desc,
        "blocks": project_blocks,
        "is_main": "main" in root
    })

project_list.sort(key=lambda x: (not x["is_main"], x["original_name"]))

final_projects = {}
html_visible = ""
html_hidden = ""

for i, p in enumerate(project_list):
    new_id = i + 1
    final_projects[new_id] = {
        "title": {"en": p["original_name"], "de": p["original_name"]},
        "category": {"en": p["category"], "de": p["category"]},
        "heroImage": p["hero_image"] if p["hero_image"] else "",
        "description": {"en": p["description"], "de": p["description"]},
        "role": {"en": "Designer", "de": "Designer"},
        "year": p["year"],
        "tools": p["tools"],
        "blocks": p["blocks"]
    }
    
    html_chunk = f"""
    <!-- Project {new_id} -->
    <div class="reveal-on-scroll group cursor-pointer" onclick="openModal({new_id})">
        <div class="hover-target relative aspect-[4/3] bg-gray-100 overflow-hidden mb-4">
            <img src="{p["hero_image"] or 'https://via.placeholder.com/800'}"
                class="w-full h-full object-cover transition-transform duration-700 group-hover:scale-105"
                alt="{p["original_name"]}"
                onerror="this.src='https://via.placeholder.com/800x600?text=Image+Unavailable'">
            <div class="absolute inset-0 bg-black/0 group-hover:bg-black/10 transition-colors duration-300"></div>
        </div>
        <div class="flex justify-between items-start">
            <div>
                <h3 class="text-lg font-semibold">{p["original_name"]}</h3>
                <p class="text-sm text-gray-500">{p["category"]}</p>
            </div>
        </div>
    </div>
    """
    
    if i < 3:
        html_visible += html_chunk
    else:
        html_hidden += html_chunk

with open("output.txt", "w", encoding="utf-8") as f:
    f.write("JSON_START\n")
    f.write(json.dumps(final_projects, indent=4))
    f.write("\nJSON_END\n")
    f.write("HTML_VISIBLE_START\n")
    f.write(html_visible)
    f.write("\nHTML_VISIBLE_END\n")
    f.write("HTML_HIDDEN_START\n")
    f.write(html_hidden)
    f.write("\nHTML_HIDDEN_END\n")
